src/common/crypto/cryptoutils.cpp

```diff
@@ -242,6 +242,44 @@ Error FindUrlParam(const String& url, const String& paramName, String& paramValu
     return paramValue.Insert(paramValue.end(), url.Get() + start, url.Get() + end);
 }
 
+static Error ParsePIN(const String& url, String& pin)
+{
+    auto pinValueErr = FindUrlParam(url, "pin-value", pin);
+    if (!pinValueErr.IsNone() && !pinValueErr.Is(ErrorEnum::eNotFound)) {
+        return AOS_ERROR_WRAP(ErrorEnum::eInvalidArgument);
+    }
+
+    StaticString<cFilePathLen> pinPath;
+
+    auto pinSourceErr = FindUrlParam(url, "pin-source", pinPath);
+    if (!pinSourceErr.IsNone() && !pinSourceErr.Is(ErrorEnum::eNotFound)) {
+        return AOS_ERROR_WRAP(ErrorEnum::eInvalidArgument);
+    }
+
+    if (pinValueErr.IsNone() && pinSourceErr.IsNone()) {
+        // either pin-source or pin-value should be provided
+        return AOS_ERROR_WRAP(ErrorEnum::eInvalidArgument);
+    }
+
+    if (pinValueErr.Is(ErrorEnum::eNotFound) && pinSourceErr.Is(ErrorEnum::eNotFound)) {
+        // if neither pin-value nor pin-source provided, default pin to empty string.
```

Comment:
Didn't understand comment, is there a typo?
```suggestion
        // if neither pin-value nor pin-source provided, default pin is empty string.
```

Reply:
I meant to set a pin to empty string. 
https://dictionary.cambridge.org/dictionary/english/default-to

Fixed.